### PR TITLE
Cleanup Microsoft.Extensions.Http.Resilience internals

### PR DESCRIPTION
**Pull request by martintmk**
Simplifying and dropping some unnecessary internal abstractions. 

Changes:

- Dropping some internal interfaces or turning them into abstract classes or just simple options.
- Dropping additional dependencies (Polling, NamedSingletons) making this library leaner.
- Enabling pooling of `RequestMessageSnapshot`.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4141)

### CODE REVIEW

File: src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Internal/StrategyKeyOptions.cs [hunk 1]
```diff
@@ -1,19 +1,15 @@
...
+using System;
 using System.Net.Http;
 
 namespace Microsoft.Extensions.Http.Resilience.Internal;
 
 /// <summary>
 /// The provider that returns the strategy key from the request message.
 /// </summary>
-internal interface IStrategyKeyProvider
+internal sealed class StrategyKeyOptions
```
Comment (geeknoid) [inline src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Internal/StrategyKeyOptions.cs:12]: Is this interface used in DI? If not, you could just be using Func<HttpRequestMessage, string> directly

File: src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Internal/StrategyKeyProviderHelper.cs [hunk 2]
```diff
@@ -17,25 +18,22 @@ public static void SelectStrategyByAuthority(IServiceCollection services, string
         {
             var redactor = serviceProvider.GetRequiredService<IRedactorProvider>().GetRedactor(classification);
 
-            return new ByAuthorityStrategyKeyProvider(redactor);
+            return new ByAuthorityStrategyKeyProvider(redactor).GetStrategyKey;
```
Comment (geeknoid) [inline src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Internal/StrategyKeyProviderHelper.cs:21]: Perhaps rename GetStrategyKey to just StrategyKey?

**Review by geeknoid — APPROVED** [review]

File: src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Internal/StrategyKeyOptions.cs (hunk 1 above)
Comment (martintmk) [reply src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Internal/StrategyKeyOptions.cs:12]: Yes, it comes from DI. This options class is just a holder for that delegate.

File: src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Internal/StrategyKeyProviderHelper.cs (hunk 2 above)
Comment (martintmk) [reply src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Internal/StrategyKeyProviderHelper.cs:21]: `GetStrategyKey` is just a function that gets converted to a delegate. I think the name is ok here

**Pull request merged**